src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,711 @@
+        let record_id_for_this_row_depth = record_id_for_each_depth[i + 1]; // skip row 0
+
+        let capped_attribution_outputs = compute_row_with_previous(
+            ctx_for_this_row_depth,
+            record_id_for_this_row_depth,
+            row,
+            &mut prev_row_inputs,
+            num_saturating_sum_bits,
+        )
+        .await?;
+
+        output.push(capped_attribution_outputs);
+    }
+
+    Ok(output)
+}
+
+///
+/// Upon encountering the first row of data from a new user (as distinguished by a different OPRF of the match key)
+/// this function encapsulates the variables that must be initialized. No communication is required for this first row.
+///
+fn initialize_new_device_attribution_variables<BK, TV>(
+    share_of_one: Replicated<Gf2>,
+    input_row: &PrfShardedIpaInputRow<BK, TV>,
+    num_saturating_sum_bits: usize,
+) -> InputsRequiredFromPrevRow
+where
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    InputsRequiredFromPrevRow {
+        ever_encountered_a_source_event: share_of_one - &input_row.is_trigger_bit,
+        attributed_breakdown_key_bits: BitDecomposed::decompose(BK::BITS, |i| {
+            input_row.breakdown_key.map(|v| Gf2::truncate_from(v[i]))
+        }),
+        saturating_sum: SaturatingSum::new(
+            BitDecomposed::new(vec![Replicated::ZERO; num_saturating_sum_bits]),
+            Replicated::ZERO,
+        ),
+        // This is incorrect in the case that the CAP is less than the maximum value of "trigger value" for a single row
+        // Not a problem if you assume that's an invalid input
+        difference_to_cap: BitDecomposed::new(vec![Replicated::ZERO; TV::BITS as usize]),
+    }
+}
+
+///
+/// To support "Last Touch Attribution" we move the `breakdown_key` of the most recent source event
+/// down to all of trigger events that follow it.
+///
+/// The logic here is extremely simple. For each row:
+/// (a) if it is a source event, take the breakdown key bits.
+/// (b) if it is a trigger event, take the breakdown key bits from the preceding line
+///
+async fn breakdown_key_of_most_recent_source_event<C>(
+    ctx: C,
+    record_id: RecordId,
+    is_trigger_bit: &Replicated<Gf2>,
+    prev_row_breakdown_key_bits: &BitDecomposed<Replicated<Gf2>>,
+    cur_row_breakdown_key_bits: &BitDecomposed<Replicated<Gf2>>,
+) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    Ok(BitDecomposed::new(
+        ctx.parallel_join(
+            cur_row_breakdown_key_bits
+                .iter()
+                .zip(prev_row_breakdown_key_bits.iter())
+                .enumerate()
+                .map(|(i, (cur_bit, prev_bit))| {
+                    let c = ctx.narrow(&BitOpStep::from(i));
+                    async move {
+                        let maybe_diff = is_trigger_bit
+                            .multiply(&(prev_bit.clone() - cur_bit), c, record_id)
+                            .await?;
```

Comment:
Isn't this prime for our if_else() protocol?

Reply:
Yes, I can use that in a few places.